lib/sqlitex/query.ex

```
@@ -1,87 +1,48 @@
-    case :esqlite3.column_names(statement) do
-      {:error, _}=other -> other
-      names -> {:ok, %Sqlitex.Query{query | column_names: names}}
-    end
-  end
+  * `db` - A sqlite database.
+  * `sql` - The query to run as a string.
+  * `opts` - Options to pass into the query.  See below for details.
 
-  defp column_types({:ok, %Sqlitex.Query{statement: statement}=query}) do
-    case :esqlite3.column_types(statement) do
-      {:error, _}=other -> other
-      types -> {:ok, %Sqlitex.Query{query | column_types: types}}
-    end
-  end
+  ## Options
 
-  defp datetime_to_string({{yr, mo, da}, {hr, mi, se, usecs}}) do
-    [zero_pad(yr, 4), "-", zero_pad(mo, 2), "-", zero_pad(da, 2), " ", zero_pad(hr, 2), ":", zero_pad(mi, 2), ":", zero_pad(se, 2), ".", zero_pad(usecs, 6)]
-    |> Enum.join
-  end
+  * `bind` - If your query has parameters in it, you should provide the options
+    to bind as a list.
+  * `into` - The collection to put results into.  This defaults to a list.
 
-  defp fetch_data({:ok, %Sqlitex.Query{statement: statement}=query}) do
-    case :esqlite3.fetchall(statement) do
-      {:error,_}=other -> other
-      raw_data -> {:ok, %Sqlitex.Query{query | raw_data: raw_data}}
-    end
+  ## Returns
+  * {:ok, results} on success
+  * {:error, _} on failure.
+  """
+  def query(db, sql, opts \\ []) do
+    pipe_with &pipe_ok/2,
+      Statement.prepare(db, sql)
+      |> Statement.bind_values(Dict.get(opts, :bind, []))
+      |> Statement.fetch_all(Dict.get(opts, :into, []))
   end
 
-  defp into_from_opts(opts), do: Keyword.get(opts, :into, [])
+  @doc """
+  Same as `query/3` but raises an error on error.
 
-  defp prepare({:ok, %Sqlitex.Query{sql: sql, database: database}=query}) do
-    case :esqlite3.prepare(sql, database) do
-      {:ok, statement} -> {:ok, %Sqlitex.Query{query | statement: statement}}
-      other -> other
-    end
+  Returns the results otherwise.
+  """
+  def query!(db, sql, opts \\ []) do
```

Currently this `query!/3` function will raise a `BadMatchError` if something goes wrong. If we are going to implement a `!` version of this function I think it should probably raise an error which is specific to the use case.

I'm not sure when I would use this function so I'm wondering if we should add it or not. Anyone have ideas of when this function is useful?